Canary docs (#5363)
solo-io/gloo: docs/content/operations/upgrading/canary.md

---
title: Canary Upgrade (1.9.0+)
weight: 5
description: Upgrading Gloo Edge with a canary workflow
---

In this guide we will describe the necessary steps to upgrade your Gloo Edge or Gloo Edge Enterprise deployments using

> a canary model. This guide assumes the older Gloo Edge version is at least 1.9.0.

Reviewer: is this right? should it be the newer version is at least 1.9.0?

Author: this is right, this doc is for the canary upgrade. to ensure `glooctl check` works and reports statuses appropriately for each controller we will need support for the multi-status that @sam-heilbron has been working on. that feature work won't land until 1.9.0

> In this guide we will describe the necessary steps to upgrade your Gloo Edge or Gloo Edge Enterprise deployments using
> a canary model. This guide assumes the older Gloo Edge version is at least 1.9.0.
> 

Reviewer 2: Maybe add a brief description of what canary upgrade is and why they would want to use it?
```suggestion
In version 1.9.0 or later, you can upgrade your Gloo Edge or Gloo Edge Enterprise deployments with
a canary model. In the canary model, you have two different `gloo` deployments in your data plane and can check that the deployment at the latest version handles traffic as you expect before upgrading to run at the latest version.

```

{{% notice note %}}

> In versions prior to 1.9.0, status reporting on Gloo CRs was not per namespace, thus we had little insight into the
> state of a resource from the perspective of each canary installation. Further, the helm value overrides for the xds
> service address and xds service port had not yet been added, e.g.:

Reviewer 2: ```suggestion
**Why can't I perform canary upgrades for versions older than 1.9?** Prior to 1.9.0, status reporting on Gloo CRs was not per namespace, so you could not monitor the state of separate Gloo resources at different versions in different namespaces as part of a canary upgrade. Also, the Helm configuration could not override the xDS service address or xDS service port, such as follow. 
```

```yaml
gatewayProxies:
  gatewayProxy: # do the following for each gateway proxy
    xdsServiceAddress: xds-relay.default.svc.cluster.local
    xdsServicePort: 9991
```

> In versions prior to 1.8.0, there was a [bug](https://github.com/solo-io/gloo/issues/5030) that prevented canary deploys
> from working as the old control plane would crash loop when it saw newly added fields.

Reviewer 2: ```suggestion
Prior to 1.8.0, a [bug](https://github.com/solo-io/gloo/issues/5030) prevented canary deployments
from working because the old control plane crash went into a crash loop when new fields were added.
```

> {{% /notice %}}

Reviewer 2: Possibly move this entire note in the prereq section after giving the required version

> ##### Prereqs

Reviewer 2: ```suggestion
## Prerequisites
```

Gloo Edge 1.9.0 or later installed.

> ##### Simple Canary (Recommended)

Reviewer 2: ```suggestion
## Simple canary upgrades (recommended approach)
```

Reviewer 2: These look like ordered steps, so my suggestions will order them.

> - **Install**
>   - Install Gloo Edge with the new version to another namespace, e.g. `glooctl install gateway --version 1.9.0 -n gloo-system-canary`.

Reviewer 2: I think the sample version would be at least 1.9.1, right, because they have to have at least 1.9.0 so that would not be a newer version.
```suggestion
1. Install the newer version of Gloo Edge in another namespace in your data plane cluster, such as with the following command.
    ```shell
     glooctl install gateway --version 1.9.1 -n gloo-system-canary
     ```
```

> - **Test**
>   - Test your routes, monitor metrics, and run `glooctl check` until you are happy with the installation.

Reviewer 2: ```suggestion
2. Test your routes and monitor the metrics of the newer version.
    ```shell
    glooctl check
    ```
```

- **Validate**

>   - `gloooctl uninstall -n gloo-system`, or `helm delete` the original installation.

Reviewer 2: Should they upgrade the deployment in the `gloo-system` namespace and then delete the `gloo-system-canary` namespace instead? Or is it fine to no longer have gloo in `gloo-system` and just have this new `-canary` ns going forward? (and then in the future for the next version, I guess they'd have `-canary2` or something?)

Reviewer 2: ```suggestion
3. Remove the older version of Gloo Edge so that your cluster uses the newer version going forward.
    With `glooctl`:
    ```shell
    gloooctl uninstall -n gloo-system
    ```
    With Helm:
    ```shell
    helm delete
    ```
```

Author: i decided to update the name to include the version instead of the `-canary` suffix, since we don't intend to clean up the second install or rename it

Reviewer 2: Sounds good, thanks!

> ##### Appendix: In-Place "Canary" using XDS-Relay

Reviewer 2: ```suggestion
## Appendix: In-place canary upgrades by using xDS relay
```

> A simple way to decouple the lifecycle of your control plane from your data plane (installed together by default
> with Gloo Edge and Gloo Edge Enterprise) is to use the [xds-relay]({{< versioned_link_path fromRoot="/operations/production_deployment/#xds-relay" >}})
> project as your "canary" control plane. This provides extra resiliency for your live xds configuration in the event of
> failure during an in-place `helm upgrade`. 

Reviewer 2: I wasn't sure if this meant that xds-relay is operating like a control plane for the canary deployment, or if this is somehow acting like a newer version to test, but I revised it as if it were the first one.
```suggestion
By default, your Gloo Edge or Gloo Edge Enterprise control plane and data plane are installed together. However, you can decouple the control and data planes by using the [`xds-relay`]({{< versioned_link_path fromRoot="/operations/production_deployment/#xds-relay" >}})
project as the control plane for the newer version deployment in a canary upgrade. This setup provides extra resiliency for your live xDS configuration in the event of failure during an in-place `helm upgrade`. 
```

Author: it's moreso the former, acting as an intermediate "control plane" (it's not smart, just providing resiliency). Updated with your reccomenedation.
